docs/site/Dynamic-models-repositories-controllers.md

```diff
@@ -0,0 +1,218 @@
+
+An example of creating a `ModelDefinition` object:
+
+```ts
+const bookDef = new ModelDefinition({
+  name: 'book',
+  properties: {
+    id: {
+      type: 'Number',
+      required: true,
+      length: null,
+      precision: 10,
+      scale: 0,
+      id: 1,
+      mysql: {
+        columnName: 'id',
+        dataType: 'int',
+        dataLength: null,
+        dataPrecision: 10,
+        dataScale: 0,
+        nullable: 'N',
+      },
+    },
+    title: {
+      type: 'String',
+      required: false,
+      length: 512,
+      precision: null,
+      scale: null,
+      mysql: {
+        columnName: 'title',
+        dataType: 'varchar',
+        dataLength: 512,
+        dataPrecision: null,
+        dataScale: null,
+        nullable: 'Y',
+      },
+    }
+  },
+  settings: {
+    idInjection: false,
+    mysql: {schema: 'test', table: 'BookStore'},
+  },
+});
+```
+
+### Defining a Model
+
+A LoopBack model class is created by passing a `ModelDefinition` object to
+`@loopback/repository`'s helper function `defineModelClass()`. It expects a
+base model to extend (typically `Model` or `Entity`), folowed by the model
+definition object. In this case it will be `Entity`.
+
+```ts
+const BookModel = defineModelClass(Entity, bookDef);
+```
+
+In case you need to use an existing Model as the base class, you should use the
+`DynamicModelCtor` class, instead of the `defineModelClass()` function.
+
+```ts
+import DynamicModelCtor from '@loopback/repository';
+// Assuming User is a pre-existing Model class in the app
+const StudentModel = new DynamicModelCtor(User, studentDef);
+```
+
+For details about `ModelDefinition`, `defineModelClass`, `Model`, `Entity`, and
+`DynamicModelCtor` refer to the
+[@loopback/repository API documentation](https://loopback.io/doc/en/lb4/apidocs.repository.html).
+
+### Defining a Datasource
+
+Before a repository for this model can be set up, a datasource should be ready.
+If required, set one up dynamically by creating an instance of
+`juggler.DataSource`. It requires a name for the datasource, the connector, and
+a connection url as shown below.
+
+```ts
+const dsName = 'bookstore-ds';
+const bookDs = new juggler.DataSource({
+  name: dsName,
+  connector: require('loopback-connector-mongodb'),
+  url: 'mongodb://sysop:moon@localhost',
+});
+await bookDs.connect();
+app.dataSource(bookDs, dsName);
+```
+
+Connect the datsource by calling its `.connect()` method and attach it to the
+app using `app.dataSource()` method. Note, this method will be available only
+on `RepositoryMixin` apps.
+
+### Defining a Repository
+
+Now that a datasource is ready, a repository can be created for `BookModel`.
+
+The `@loopback/rest-crud` package exports a helper function,
+`defineCrudRepositoryClass`, this function creates a
+`defineCrudRepositoryClass`-based entity CRUD repository class for the given
+model.
+
+```ts
+const BookRepository = defineCrudRepositoryClass(BookModel);
+```
```

Comment:
This will create a repository with `id` typed as `unknown` and empty relations (no navigational properties). Again, this may be ok if users don't need to access `findById` or included related models from code. However, in many cases, users will need the repository class to understand the real id type and the actual navigational properties. We should show them how to accomplish that:

```ts
const BookRepository = defineCrudRepositoryClass<
  Book,
  typeof Book.prototype.id,
  BookRelations
>(BookModel);
```